File: examples/adt.lua
Just = function(x)
  return function(handler)
    return handler.Just(x)
  end
end

Nothing = function(handler)
  return handler.Nothing
end

function match(value, handler)
  return value(handler)
end

function safeDivide(numerator, denominator)
  if denominator == 0
    then return Nothing
    else return Just(numerator / denominator)
  end
end

function displayResult(maybeValue)
  extractedValue = match(maybeValue, {
    Just = function(result) return "Result of division is " .. result end,
    Nothing = "Handled divide by zero error"
  })
  print(extractedValue)
end

badDivision = safeDivide(5, 0)
goodDivision = safeDivide(22, 7)

displayResult(badDivision)
displayResult(goodDivision)

File: examples/adt.clj
(def Just (fn [x] (fn [handler] ((:Just handler) x))))
(def Nothing (fn [handler] (:Nothing handler)))

(defn match [value handler] (value handler))

(defn safeDivide [numerator, denominator]
  (if (= 0 denominator) Nothing (Just (/ numerator denominator))))

(defn displayResult [maybeValue]
  (let [extractedValue (match maybeValue {
    :Just (fn [result] (str "Result of division is " result))
    :Nothing "Handled divide by zero error"})]
    (println extractedValue)))

(def badDivision (safeDivide 5 0))
(def goodDivision (safeDivide 22 7))

(displayResult badDivision)
(displayResult goodDivision)
